### PR TITLE
feat: add `filter` method to `array/complex64`

### PR DESCRIPTION
**Pull request by Jaysukh-409**
## Description

> What is the purpose of this pull request?

This pull request:

-  Adds the typed array `filter` method to prototype of `Complex64Array`

## Related Issues

> Does this pull request have any related issues?

No.

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

File: lib/node_modules/@stdlib/array/complex64/README.md [hunk 1]
```diff
@@ -856,6 +856,72 @@ var count = context.count;
 // returns 3
 ```
 
+<a name="method-filter"></a>
+
+#### Complex64Array.prototype.filter( predicate\[, thisArg])
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/complex64/README.md]:
```suggestion
#### Complex64Array.prototype.filter( predicate\[, thisArg] )
```

File: lib/node_modules/@stdlib/array/complex64/README.md [hunk 2]
```diff
@@ -856,6 +856,72 @@ var count = context.count;
...
+
+Returns the array containing the elements of an array which pass a test implemented by a predicate function.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/complex64/README.md]:
```suggestion
Returns a new array containing the elements of an array which pass a test implemented by a predicate function.
```

File: lib/node_modules/@stdlib/array/complex64/README.md [hunk 3]
```diff
@@ -856,6 +856,72 @@ var count = context.count;
...
+// Set the first three elements:
+arr.set( [ 1.0, -1.0 ], 0 );
+arr.set( [ 2.0, 2.0 ], 1 );
+arr.set( [ 3.0, -3.0 ], 2 );
+
+var out = arr.filter( predicate );
+// returns <Complex64Array>
+
+var len = out.length;
+// returns 1
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/complex64/README.md:884]: @Jaysukh-409 I suggest extending this example to actually show what the first, and only, element is for the returned array. You can use `get`, followed by `realf` and `imagf`.

File: lib/node_modules/@stdlib/array/complex64/docs/types/index.d.ts [hunk 4]
```diff
@@ -343,6 +343,35 @@ declare class Complex64Array implements Complex64ArrayInterface {
 	*/
 	every<U = unknown>( predicate: Predicate, thisArg?: ThisParameterType<Predicate<U>> ): boolean;
 
+	/**
+	* Returns the array containing the elements of an array which pass a test implemented by a predicate function.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/complex64/docs/types/index.d.ts]:
```suggestion
	* Returns a new array containing the elements of an array which pass a test implemented by a predicate function.
```

File: lib/node_modules/@stdlib/array/complex64/docs/types/index.d.ts [hunk 5]
```diff
@@ -343,6 +343,35 @@ declare class Complex64Array implements Complex64ArrayInterface {
...
+	*
+	* @param predicate - test function
+	* @param thisArg - execution context
+	* @returns new array containing elements passing test
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/complex64/docs/types/index.d.ts]:
```suggestion
	* @returns new array containing elements which pass a test implemented by a predicate function
```

File: lib/node_modules/@stdlib/array/complex64/docs/types/index.d.ts [hunk 6]
```diff
@@ -343,6 +343,35 @@ declare class Complex64Array implements Complex64ArrayInterface {
...
+	*
+	* arr.set( [ 1.0, -1.0 ], 0 );
+	* arr.set( [ 2.0, 2.0 ], 1 );
+	* arr.set( [ 3.0, -3.0 ], 2 );
+	*
+	* var out = arr.filter( predicate );
+	* // returns <Complex64Array>
+	*
+	* var len = out.length;
+	* // returns 1
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/complex64/docs/types/index.d.ts:371]: @Jaysukh-409 Same comment as in the README.

File: lib/node_modules/@stdlib/array/complex64/lib/main.js [hunk 7]
```diff
@@ -919,6 +919,58 @@ setReadOnly( Complex64Array.prototype, 'every', function every( predicate, thisA
 	return true;
 });
 
+/**
+* Returns the array containing the elements of an array which pass a test implemented by a predicate function.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/complex64/lib/main.js]:
```suggestion
* Returns a new array containing the elements of an array which pass a test implemented by a predicate function.
```

File: lib/node_modules/@stdlib/array/complex64/lib/main.js [hunk 8]
```diff
@@ -919,6 +919,58 @@ setReadOnly( Complex64Array.prototype, 'every', function every( predicate, thisA
...
+*
+* arr.set( [ 1.0, -1.0 ], 0 );
+* arr.set( [ 2.0, 2.0 ], 1 );
+* arr.set( [ 3.0, -3.0 ], 2 );
+*
+* var out = arr.filter( predicate );
+* // returns <Complex64Array>
+*
+* var len = out.length;
+* // returns 1
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/complex64/lib/main.js:952]: @Jaysukh-409 Same comment as in README.

File: lib/node_modules/@stdlib/array/complex64/lib/main.js [hunk 9]
```diff
@@ -919,6 +919,58 @@ setReadOnly( Complex64Array.prototype, 'every', function every( predicate, thisA
...
+		throw new TypeError( 'invalid invocation. `this` is not a complex number array.' );
+	}
+	if ( !isFunction( predicate ) ) {
+		throw new TypeError( format( 'invalid argument. First argument must be a function. Value: `%s`.', predicate ) );
+	}
+	buf = this._buffer;
+	out = [];
+	for ( i = 0; i < this._length; i++ ) {
+		if ( predicate.call( thisArg, getComplex64( buf, i ), i, this ) ) {
+			out.push( getComplex64( buf, i ) );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/complex64/lib/main.js]: I would avoid duplicate invocation of `getComplex64`. Instead, assign to an intermediate variable `z` prior to calling the `predicate` function.

File: lib/node_modules/@stdlib/array/complex64/README.md [hunk 10]
```diff
@@ -856,6 +856,82 @@ var count = context.count;
...
+<a name="method-filter"></a>
+
+#### Complex64Array.prototype.filter( predicate\[, thisArg] )
+
+Returns a new array containing the elements of an array which pass a test implemented by a predicate function.
+
+```javascript
+var realf = require( '@stdlib/complex/realf' );
+var imagf = require( '@stdlib/complex/imagf' );
+var Complex64 = require( '@stdlib/complex/float32' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/complex64/README.md]: @Jaysukh-409 This doesn't appear to be used, correct?

File: lib/node_modules/@stdlib/array/complex64/docs/types/index.d.ts [hunk 11]
```diff
@@ -343,6 +343,45 @@ declare class Complex64Array implements Complex64ArrayInterface {
...
+	* Returns a new array containing the elements of an array which pass a test implemented by a predicate function.
+	*
+	* @param predicate - test function
+	* @param thisArg - execution context
+	* @returns new array containing elements which pass a test implemented by a predicate function
+	*
+	* @example
+	* var realf = require( '@stdlib/complex/realf' );
+	* var imagf = require( '@stdlib/complex/imagf' );
+	* var Complex64 = require( '@stdlib/complex/float32' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/complex64/docs/types/index.d.ts]: @Jaysukh-409 Same comment as README.

File: lib/node_modules/@stdlib/array/complex64/lib/main.js [hunk 12]
```diff
@@ -919,6 +919,70 @@ setReadOnly( Complex64Array.prototype, 'every', function every( predicate, thisA
...
+* @param {Function} predicate - test function
+* @param {*} [thisArg] - predicate function execution context
+* @throws {TypeError} `this` must be a complex number array
+* @throws {TypeError} first argument must be a function
+* @returns {Complex64Array} complex number array
+*
+* @example
+* var realf = require( '@stdlib/complex/realf' );
+* var imagf = require( '@stdlib/complex/imagf' );
+* var Complex64 = require( '@stdlib/complex/float32' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/complex64/lib/main.js]: @Jaysukh-409 Same comment as README.

File: lib/node_modules/@stdlib/array/complex64/README.md (hunk 10 above)
Comment (Jaysukh-409) [reply lib/node_modules/@stdlib/array/complex64/README.md]: Yes, Sorry for that. We don't need the Complex64 here.

File: lib/node_modules/@stdlib/array/complex64/test/test.filter.js [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+
+// TESTS //
+
+tape( 'main export is a function', function test( t ) {
+	t.ok( true, __filename );
+	t.strictEqual( typeof Complex64Array, 'function', 'main export is a function' );
+	t.end();
+});
+
+tape( 'attached to the prototype of the main export is an `filter` method', function test( t ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/complex64/test/test.filter.js]:
```suggestion
tape( 'attached to the prototype of the main export is a `filter` method', function test( t ) {
```

File: lib/node_modules/@stdlib/array/complex64/test/test.filter.js [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+tape( 'the method returns a new complex number array containing only those elements which satisfy a test condition', function test( t ) {
+	var expected;
+	var actual;
+	var arr;
+
+	arr = new Complex64Array( [ 1.0, -1.0, 2.0, 2.0, 3.0, -3.0 ] );
+	expected = new Complex64Array( [ 1.0, -1.0, 3.0, -3.0 ] );
+	actual = arr.filter( predicate );
+
+	t.deepEqual( actual, expected, 'returns expected value' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/complex64/test/test.filter.js]: @Jaysukh-409 Unfortunately, I don't think this works the way you want it to. Instead, we want to reinterpret the `Complex64Array` to a `Float32Array` (see, e.g., https://github.com/stdlib-js/stdlib/blob/1464d5bac230c75ec5c98c59ad36c5e58d7d0e21/lib/node_modules/%40stdlib/array/full/test/test.js#L204) and then deep equal the reinterpreted array to an expected `Float32Array`.

File: lib/node_modules/@stdlib/array/complex64/test/test.filter.js [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+tape( 'the method returns same array if all elements satisfy a test condition', function test( t ) {
+	var expected;
+	var actual;
+	var arr;
+
+	arr = new Complex64Array( [ 1.0, -1.0, 2.0, -2.0, 3.0, -3.0 ] );
+	expected = arr;
+	actual = arr.filter( predicate );
+
+	t.deepEqual( actual, expected, 'returns expected value' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/complex64/test/test.filter.js]: @Jaysukh-409 Same comment as above.

Also, the method doesn't return the "same" array, but a copy of the array having the same elements.

Comment (kgryte) [reply lib/node_modules/@stdlib/array/complex64/test/test.filter.js]: In which case, we want to assert that

```
t.notEqual( actual, arr, 'returns new instance' );
```

File: lib/node_modules/@stdlib/array/complex64/test/test.filter.js [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+	var ctx;
+
+	arr = new Complex64Array( [ 1.0, -1.0, 2.0, 2.0, 3.0, -3.0 ] );
+	expected = new Complex64Array( [ 1.0, -1.0, 3.0, -3.0 ] );
+	ctx = {
+		'count': 0
+	};
+	actual = arr.filter( predicate, ctx );
+
+	t.deepEqual( actual, expected, 'returns expected value' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/complex64/test/test.filter.js]: Same comment as above.

File: lib/node_modules/@stdlib/array/complex64/README.md (hunk 10 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/array/complex64/README.md]: @Jaysukh-409 No worries. If you could remove here and in other examples, that would be great!

File: lib/node_modules/@stdlib/array/complex64/test/test.filter.js [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+	t.strictEqual( actual.length, expected.length/2, 'returns expected value' );
+	t.deepEqual( reinterpret64( actual, 0 ), expected, 'returns expected value' );
+	t.end();
+
+	function predicate( v ) {
+		return ( realf( v ) > 0 && imagf( v ) < 0 );
+	}
+});
+
+tape( 'the method returns same array if all elements satisfy a test condition', function test( t ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/complex64/test/test.filter.js]:
```suggestion
tape( 'the method copies all elements to a new array if all elements satisfy a test condition', function test( t ) {
```

File: lib/node_modules/@stdlib/array/complex64/test/test.filter.js [hunk 18]
```diff
@@ -0,0 +1,189 @@
...
+	var expected;
+	var actual;
+	var arr;
+
+	arr = new Complex64Array( [ 1.0, -1.0, 2.0, -2.0, 3.0, -3.0 ] );
+	expected = new Float32Array( [ 1.0, -1.0, 2.0, -2.0, 3.0, -3.0 ] );
+	actual = arr.filter( predicate );
+
+	t.strictEqual( instanceOf( actual, Complex64Array ), true, 'returns expected value' );
+	t.strictEqual( actual.length, expected.length/2, 'returns expected value' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/array/complex64/test/test.filter.js:158]:
```suggestion
	t.notEqual( actual, arr, 'returns a new instance' );
	t.strictEqual( actual.length, expected.length/2, 'returns expected value' );
```

**Review by kgryte — APPROVED** [review]
LGTM. Thanks, @Jaysukh-409!

**Pull request merged**